Mathlib/RingTheory/Bialgebra/TensorProduct.lean

```diff
@@ -14,126 +14,110 @@ We define the data in the monoidal structure on the category of bialgebras - e.g
-
-Since the coalgebra instance on a tensor product of coalgebras is defined using a universe
-monomorphic monoidal category structure on `ModuleCat R`, we require our bialgebras to be in the
-same universe as the base ring `R` here too. Any contribution that achieves universe polymorphism
-would be welcome. For instance, the tensor product of bialgebras in the
-[FLT repo](https://github.com/ImperialCollegeLondon/FLT/blob/eef74b4538c8852363936dfaad23e6ffba72eca5/FLT/mathlibExperiments/Coalgebra/TensorProduct.lean)
-is already universe polymorphic since it does not go via category theory.
-
 -/
 
-universe v u
-
 open scoped TensorProduct
 
 namespace Bialgebra.TensorProduct
 
-variable (R A B : Type*) [CommRing R] [Ring A] [Ring B] [Bialgebra R A] [Bialgebra R B]
+open Coalgebra.TensorProduct
+
+variable (R A B C D : Type*) [CommSemiring R] [Semiring A] [Semiring B]
+  [Bialgebra R A] [Bialgebra R B]
 
 lemma counit_eq_algHom_toLinearMap :
     Coalgebra.counit (R := R) (A := A ⊗[R] B) =
-      ((Algebra.TensorProduct.lmul' R).comp (Algebra.TensorProduct.map
-      (Bialgebra.counitAlgHom R A) (Bialgebra.counitAlgHom R B))).toLinearMap := by
+      ((Algebra.TensorProduct.rid _ _ _).toAlgHom.comp (Algebra.TensorProduct.map
+      (Bialgebra.counitAlgHom R A) (Bialgebra.counitAlgHom R B))).toLinearMap :=
   rfl
 
 lemma comul_eq_algHom_toLinearMap :
     Coalgebra.comul (R := R) (A := A ⊗[R] B) =
       ((Algebra.TensorProduct.tensorTensorTensorComm R R A A B B).toAlgHom.comp
       (Algebra.TensorProduct.map (Bialgebra.comulAlgHom R A)
-      (Bialgebra.comulAlgHom R B))).toLinearMap := by
-  ext; simp [← Algebra.TensorProduct.toLinearEquiv_tensorTensorTensorComm]
-
-end Bialgebra.TensorProduct
+      (Bialgebra.comulAlgHom R B))).toLinearMap :=
+  rfl
 
-noncomputable instance TensorProduct.instBialgebra
-    {R A B : Type u} [CommRing R] [Ring A] [Ring B]
-    [Bialgebra R A] [Bialgebra R B] : Bialgebra R (A ⊗[R] B) := by
-  have hcounit := congr(DFunLike.coe $(Bialgebra.TensorProduct.counit_eq_algHom_toLinearMap R A B))
-  have hcomul := congr(DFunLike.coe $(Bialgebra.TensorProduct.comul_eq_algHom_toLinearMap R A B))
+noncomputable instance _root_.TensorProduct.instBialgebra : Bialgebra R (A ⊗[R] B) := by
+  have hcounit := congr(DFunLike.coe $(counit_eq_algHom_toLinearMap R A B))
+  have hcomul := congr(DFunLike.coe $(comul_eq_algHom_toLinearMap R A B))
   refine Bialgebra.mk' R (A ⊗[R] B) ?_ (fun {x y} => ?_) ?_ (fun {x y} => ?_) <;>
   simp_all only [AlgHom.toLinearMap_apply] <;>
   simp only [map_one, map_mul]
 
-namespace Bialgebra.TensorProduct
+variable {R A B C D}
 
-variable {R A B C D : Type u} [CommRing R] [Ring A] [Ring B] [Ring C] [Ring D]
-    [Bialgebra R A] [Bialgebra R B] [Bialgebra R C] [Bialgebra R D]
+variable [Semiring C] [Semiring D] [Bialgebra R C] [Bialgebra R D]
 
 /-- The tensor product of two bialgebra morphisms as a bialgebra morphism. -/
-noncomputable def map (f : A →ₐc[R] B) (g : C →ₐc[R] D) :
```

Comment:
What's the reason for the variable swap here? Does it make sense to split to its own PR?

Reply:
This PR is split from #25171 and I need to change the variables around there because different bialgebras are over different bases in that PR. I hope this is fine as long as I'm not making mathlib worse. But I can revert the changes here and redo them in that PR if you prefer.